compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -3144,6 +3144,127 @@ static inline void loadArrayCmpSources(TR::Node *node, TR::InstOpCode::Mnemonic
+   // generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rldicr, node, temp2Reg, returnReg, 56, 0xFF00000000000000);   
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src1AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec0Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, tempReg, src2AddrReg, indexReg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lxvll, node, vec1Reg, tempReg, temp2Reg);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vcmpequb, node, vec0Reg, vec0Reg, vec1Reg);
+   // residue end
+
+   // result
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, resultLabel);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vnor, node, vec0Reg, vec0Reg, vec0Reg);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::vclzlsbb, node, tempReg, vec0Reg);
+
+   if (!node->isArrayCmpLen())
+      {
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, returnReg, returnReg, -1);
+      }
+
+   // offset = matched-byte-count >= remainder ? remainder : match-byte-count
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, condReg, tempReg, returnReg);
+   generateTrg1Src3Instruction(cg, TR::InstOpCode::isellt, node, returnReg, tempReg, returnReg, condReg);
+
+   // index = index + offset, if we need to return unmatched index, then we are done here 
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, indexReg, returnReg);
+   
+   if (!node->isArrayCmpLen())
+      {
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::lbzx, node, tempReg, src1AddrReg, returnReg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::lbzx, node, indexReg, src2AddrReg, returnReg);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::cmp4, node, condReg, tempReg, indexReg);
+      // result = -1,0,1
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::setb, node, tempReg, condReg);
+      // convert -1,0,1 to 1,0,2 to match current arraycmp return value
+      generateTrg1Src1Instruction(cg, TR::InstOpCode::neg, node, tempReg, tempReg);
+      generateTrg1Src1Imm2Instruction(cg, TR::InstOpCode::rlwinm, node, returnReg, tempReg, 2, 3);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, returnReg, returnReg, tempReg);
+      }
+
+   int32_t numRegs = 9;
+
+   TR::RegisterDependencyConditions *dependencies = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, numRegs, cg->trMemory());
+   dependencies->addPostCondition(src1AddrReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(src2AddrReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(tempReg, TR::RealRegister::NoReg);
+   dependencies->getPostConditions()->getRegisterDependency(2)->setExcludeGPR0();
+   dependencies->addPostCondition(returnReg, TR::RealRegister::NoReg);
+   if (!node->isArrayCmpLen())
+      dependencies->getPostConditions()->getRegisterDependency(3)->setExcludeGPR0();
```

Comment:
src1 and src2 registers should set excludingGPR0 too (in lbzx instructions).

also, setting excludingGPR0 unconditionally is more readable and straight-forward as well, and it doesn't incur side-effect since you have at least one register to which gr0 can be assigned.